### DISCUSSION TITLE
Pending updates included in queries before transaction committed

### DISCUSSION THREAD

**Discussion by Thomas-mcinally** [sqlalchemy/sqlalchemy#9554] [Usage Questions]
DB: Postgres

Make db_session like this:
```
    engine = create_engine("postgresql://fake:postgres@localhost:5432/fake")
    Session = sessionmaker(autocommit=False, autoflush=False, bind=engine)
    db_session = Session()
```

The below code behaves as i would expect, when i query the Application table i only get changes which have been committed to database (i.e. only read what is in the database)
```
db_session.add(Application(user_id=111, status="rejected"))
db_session.query(Application).all() #expect no entry because haven't committed yet
>> []
db_session.commit()
db_session.query(Application).all() #expect to see new entry because have committed
>> [<models.application.application.Application object at 0x125ddf2b0>]
```

The below code is unexpected, however. 
```
stmt = (
            update(Application)
            .where(Application.user_id == 111)
            .values(status="approved")
)
db_session.execute(stmt) 

db_session.query(Application).all()[0].status
>> "approved"

db_session.rollback()
db_session.query(Application).all()[0].status
>> "rejected"
 ```

It seems that when I query the database, any pending (non-committed) updates made during the current transaction are included in the return of the query. **Is there any way i can disable this?**

**Comment by zzzeek** [comment]
you might be looking for synchronize_session=False: https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#selecting-a-synchronization-strategy 

however if that query for Application is returning Application objects that otherwise dont exist locally, then it's only going to get the SQL that you ran in your UPDATE statement.

**Reply by Thomas-mcinally** [reply]
Hi @zzzeek ,

Thanks for pointing me in the correct direction. That does indeed give the desired behaviour. 

Do you know of any way to enable this option as the default for the db_session?

**Reply by zzzeek** [reply]
we would like the `Session` to have an execution_options collection as its top level default but this has not been implemented yet.

a more verbose way to get this effect right now is to use a [listener](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute)


```py
from sqlalchemy import event

session_factory = sessionmaker(bind=engine, ...)

@event.listens_for(session_factory, 'do_orm_execute')
def receive_do_orm_execute(orm_execute_state):
    orm_execute_state.update_execution_options(synchronize_session=False)
```
